### PR TITLE
Divide the package manager into separate components

### PR DESCRIPTION
**Pull request by akshita31**
Divided the package manager into 5 components:
1. FileDownloader - Component that takes in a file path on disk and a URL, and downloads the file at that file path
2. ZipInstaller - Component that takes a path to a file(zip) on disk, and unzips it at a specified install path 
3. Package - Definition for the Package interface and PackageError class
4. PackageFilterer - Takes a list of packages and returns a list containing only the packages specific to the particular platform and also remove the already installed packages.
5. PackageManager - Invokes PackageFilterer to filter the packages and then uses the downloader and installer on each package.

A CreateTmpAsset utility has also been added to deal with creation of temporary files and folders.
Aslo to use mocking with https server, fake certificate files have been added.

Tests have been added for the same.

### CODE REVIEW

File: src/CSharpExtDownloader.ts [hunk 1]
```diff
@@ -4,68 +4,62 @@
...
-            installationStage = 'downloadPackages';
-            await this.packageManager.DownloadPackages(this.eventStream, this.proxy, this.strictSSL);
-
-            installationStage = 'installPackages';
-            await this.packageManager.InstallPackages(this.eventStream);
-
+            let runTimeDependencies = this.GetRunTimeDependenciesPackages();
+            runTimeDependencies.forEach(pkg => ResolveFilePaths(pkg));
+            installationStage = 'downloadAndInstallPackages';
+            await DownloadAndInstallPackages(runTimeDependencies, this.provider, this.platformInfo, this.eventStream);
```
[the remark this answers is not in the document]
Comment (akshita31) [reply src/CSharpExtDownloader.ts]: We are planning to remove the Lockfile concept entirely and using the InstallTestPath to check whether the packages are installed or not.

File: src/CSharpExtDownloader.ts [hunk 2]
```diff
@@ -4,68 +4,62 @@
...
             }
             catch (error) { }
             return success;
         }
     }
-}
+
+    private GetRunTimeDependenciesPackages(): Package[] {
+        if (this.packageJSON.runtimeDependencies) {
+            return JSON.parse(JSON.stringify(<Package[]>this.packageJSON.runtimeDependencies));
```
[the remark this answers is not in the document]
Comment (akshita31) [reply src/CSharpExtDownloader.ts]: By this we are copying these packages by 'value' using this method

File: src/CSharpExtDownloader.ts [hunk 3]
```diff
@@ -4,68 +4,62 @@
...
+        }
+
+        return null;
```
[the remark this answers is not in the document]
Comment (akshita31) [reply src/CSharpExtDownloader.ts]: Yes right. Changed it to throw an error instead. However I am not sure what the scenario would be. 
We had this check in the packages.ts , hence I used the same here - https://github.com/OmniSharp/omnisharp-vscode/pull/2188/files#diff-b28d79ec3e8b65c75f9ee73da2cf52d4L73

Do you think we should remove the check ?

File: src/CSharpExtDownloader.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (colombod) [reply src/CSharpExtDownloader.ts]: This is the common practice to "deep copy" json object in js... it is awkward but it is what it is

File: src/packageManager/PackageManager.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import { EventStream } from '../EventStream';
+import { NetworkSettingsProvider } from "../NetworkSettings";
+import { filterPackages } from "./PackageFilterer";
+import { CreateTmpFile, TmpAsset } from "../CreateTmpAsset";
+
+//Package manager needs a list of packages to be filtered based on platformInfo then download and install them
+//Note that the packages that this component will install needs absolute paths for the installPath, intsallTestPath and the binaries
+export async function DownloadAndInstallPackages(packages: Package[], provider: NetworkSettingsProvider, platformInfo: PlatformInformation, eventStream: EventStream) {
+    let filteredPackages = await filterPackages(packages, platformInfo);
+    let tmpFile: TmpAsset;
```
Comment (TheRealPiotrP) [inline src/packageManager/PackageManager.ts]: This line would be better between line 21, 22

**Comment by akshita31** [comment]
The omnisharp version has been changed for the integration tests temporarily until https://github.com/OmniSharp/omnisharp-roslyn/issues/1172 is resolved.

cc  @rchande @TheRealPiotrP

**Pull request merged**